File: java/po/world_simulator/Vector2.java
package po.world_simulator;

public class Vector2 {
    public int x;
    public int y;

    public static final Vector2 ZERO = new Vector2(0, 0);

    public Vector2(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public Vector2() {
        this.x = 0;
        this.y = 0;
    }

    public Vector2(String str) {
        int pos = str.indexOf('_');
        this.x = Integer.parseInt(str.substring(0, pos));
        this.y = Integer.parseInt(str.substring(pos + 1));
    }

    @Override
    public String toString() {
        return x + "_" + y;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Vector2 vector2 = (Vector2) obj;
        return x == vector2.x && y == vector2.y;
    }

    public Vector2 add(Vector2 other) {
        return new Vector2(this.x + other.x, this.y + other.y);
    }

    public void addSelf(Vector2 other) {
        this.x += other.x;
        this.y += other.y;
    }

    public Vector2 multiply(float scale) {
        return new Vector2((int)(this.x * scale), (int)(this.y * scale));
    }

    public void multiplySelf(int scale) {
        this.x *= scale;
        this.y *= scale;
    }
}

File: python/po/world_simulator/Vector2.py
class Vector2:
    def __init__(self, x=0, y=0): # [AI] Merged constructors and String constructor
        if isinstance(x, str):
            pos = x.find('_')
            self.x = int(x[:pos])
            self.y = int(x[pos + 1:])
        else:
            self.x = x
            self.y = y

    def __str__(self):
        return f"{self.x}_{self.y}"

    def __eq__(self, other):
        if self is other:
            return True
        if other is None or not isinstance(other, Vector2):
            return False
        return self.x == other.x and self.y == other.y

    def add(self, other):
        return Vector2(self.x + other.x, self.y + other.y)

    def addSelf(self, other):
        self.x += other.x
        self.y += other.y

    def multiply(self, scale):
        return Vector2(int(self.x * scale), int(self.y * scale))

    def multiplySelf(self, scale):
        self.x *= scale
        self.y *= scale

Vector2.ZERO = Vector2(0, 0) # [AI] Added static ZERO after class definition
